File: Assets/Typescript~/src/Client/Controllers/Match/ShopkeeperController.ts
import { Controller, OnStart } from "@easy-games/flamework-core";
import { Entity } from "Shared/Entity/Entity";
import { Network } from "Shared/Network";
import { CollectionManager } from "Shared/Util/CollectionManager";
import { CollectionTag } from "Shared/Util/CollectionTag";
import { Task } from "Shared/Util/Task";
import { ItemShopController } from "../Global/ItemShop/ItemShopController";
import { ProximityPrompt } from "../Global/ProximityPrompt/ProximityPrompt";
import { TeamUpgradeController } from "../Global/TeamUpgrade/TeamUpgradeController";

/** Proximity prompt offset. */
const PROXIMITY_PROMPT_OFFSET = new Vector3(0, 0.2, 0);

@Controller({})
export class ShopkeeperController implements OnStart {
	constructor(
		private readonly teamUpgradeController: TeamUpgradeController,
		private readonly itemShopController: ItemShopController,
	) {}

	OnStart(): void {
		Network.ServerToClient.ItemShop.AddNPCs.Client.OnServerEvent((entityIds) => {
			for (const id of entityIds) {
				Task.Spawn(async () => {
					const entity = await Entity.WaitForId(id);
					if (!entity) {
						warn("Failed to find Item Shop entity: " + id);
						return;
					}
					const prompt = new ProximityPrompt({
						promptPosition: entity.GetMiddlePosition().add(PROXIMITY_PROMPT_OFFSET),
						activationKey: KeyCode.F,
						activationKeyString: "F",
						activationRange: 3.5,
						bottomText: "Item Shop",
						topText: "Open",
					});
					/* Open shop UI on prompt activation. */
					prompt.OnActivated.Connect(() => {
						this.itemShopController.Open();
					});
				});
			}
		});
		Network.ServerToClient.TeamUpgradeShop.AddNPCs.Client.OnServerEvent((entityIds) => {
			for (const id of entityIds) {
				Task.Spawn(async () => {
					const entity = await Entity.WaitForId(id);
					if (!entity) {
						warn("Failed to find Team Upgrades entity: " + id);
						return;
					}
					const prompt = new ProximityPrompt({
						promptPosition: entity.GetMiddlePosition().add(PROXIMITY_PROMPT_OFFSET),
						activationKey: KeyCode.F,
						activationKeyString: "F",
						activationRange: 3.5,
						bottomText: "Team Upgrades",
						topText: "Open",
					});
					/* Open shop UI on prompt activation. */
					prompt.OnActivated.Connect(() => {
						this.teamUpgradeController.Open();
					});
				});
			}
		});
		/* Listen for shopkeeper creation and create proximity prompts accordingly. */
		CollectionManager.WatchCollectionTag(CollectionTag.ITEM_SHOP_SHOPKEEPER, (gameObject) => {
			print("shop keeper: ", gameObject);
			print("name: " + gameObject.name);
		});
		CollectionManager.WatchCollectionTag(CollectionTag.TEAM_UPGRADES_SHOPKEEPER, (gameObject) => {
			const prompt = new ProximityPrompt({
				promptPosition: gameObject.transform.position.add(PROXIMITY_PROMPT_OFFSET),
				activationKey: KeyCode.F,
				activationKeyString: "F",
				activationRange: 3.5,
				bottomText: "Upgrades",
				topText: "Open",
			});
			/* Open team upgrade UI on prompt activation. */
			prompt.OnActivated.Connect(() => {
				this.teamUpgradeController.Open();
			});
		});
	}
}

File: Assets/Bundles/Client/Resources/TS/Controllers/Match/ShopkeeperController.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local CollectionManager = require("Shared/TS/Util/CollectionManager").CollectionManager
local ProximityPrompt = require("Client/TS/Controllers/Global/ProximityPrompt/ProximityPrompt").ProximityPrompt
-- * Proximity prompt offset.
local PROXIMITY_PROMPT_OFFSET = Vector3.new(0, 1.5, 0)
local ShopkeeperController
do
	ShopkeeperController = setmetatable({}, {
		__tostring = function()
			return "ShopkeeperController"
		end,
	})
	ShopkeeperController.__index = ShopkeeperController
	function ShopkeeperController.new(...)
		local self = setmetatable({}, ShopkeeperController)
		return self:constructor(...) or self
	end
	function ShopkeeperController:constructor(teamUpgradeController, itemShopController)
		self.teamUpgradeController = teamUpgradeController
		self.itemShopController = itemShopController
	end
	function ShopkeeperController:OnStart()
		-- Listen for shopkeeper creation and create proximity prompts accordingly.
		CollectionManager:WatchCollectionTag("item_shop_shopkeeper", function(gameObject)
			local prompt = ProximityPrompt.new({
				promptPosition = gameObject.transform.position + PROXIMITY_PROMPT_OFFSET,
				activationKey = 20,
				activationKeyString = "F",
				activationRange = 3.5,
				bottomText = "Item Shop",
				topText = "Open",
			})
			-- Open shop UI on prompt activation.
			prompt.OnActivated:Connect(function()
				self.itemShopController:Open()
			end)
		end)
		CollectionManager:WatchCollectionTag("team_upgrades_shopkeeper", function(gameObject)
			local prompt = ProximityPrompt.new({
				promptPosition = gameObject.transform.position + PROXIMITY_PROMPT_OFFSET,
				activationKey = 20,
				activationKeyString = "F",
				activationRange = 3.5,
				bottomText = "Upgrades",
				topText = "Open",
			})
			-- Open team upgrade UI on prompt activation.
			prompt.OnActivated:Connect(function()
				self.teamUpgradeController:Open()
			end)
		end)
	end
end
-- (Flamework) ShopkeeperController metadata
Reflect.defineMetadata(ShopkeeperController, "identifier", "Bundles/Client/Controllers/Match/ShopkeeperController@ShopkeeperController")
Reflect.defineMetadata(ShopkeeperController, "flamework:parameters", { "Bundles/Client/Controllers/Global/TeamUpgrade/TeamUpgradeController@TeamUpgradeController", "Bundles/Client/Controllers/Global/ItemShop/ItemShopController@ItemShopController" })
Reflect.defineMetadata(ShopkeeperController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(ShopkeeperController, "$:flamework@Controller", Controller, { {} })
return {
	ShopkeeperController = ShopkeeperController,
}
-- ----------------------------------
-- ----------------------------------
